### DISCUSSION TITLE
RFC: `beginInclusion` with Security S2

### DISCUSSION THREAD

**Discussion by AlCalzone** [zwave-js/zwave-js#3168] [General] [answered]
I'm making great progress in supporting Security S2. Now that I'm working on device inclusion, I want to collect some feedback on how this should work.

With Security S0, things were pretty simple:
```ts
public async beginInclusion(includeNonSecure: boolean = false): Promise<boolean>
```
We use secure inclusion by default unless the application opts out with the `includeNonSecure` flag or the device doesn't support S0.

Now with S2, things are more complicated, because there is user interaction necessary, unless using SmartStart. While being included, devices request one or more security classes (i.e. different keys) and potentially client-side authentication (CSA).
This list of security classes MUST be presented to the user to choose which ones they want to grant. This includes S0, so a node included with S2 may communicate entirely with S0 afterwards.
Furthermore, the user MUST validate or enter the DSK of the device, unless "S2 Unauthenticated" or "S0" is the highest granted class. If the device requests CSA, the user MUST decide if that is okay and then interact with the device to authenticate the controller instead.

I don't think we can keep this secure-by-default behavior of the `beginInclusion` method because S2 requires these callback to be provided by the application and there needs to be a way to select the security mode (insecure, S0, S2). Furthermore, there MUST be feedback to the user if a node is included with a lower security than intended, e.g. if S0/S2 bootstrapping failed. I'll call this `InclusionResult` for now without a specific implementation.

I suppose these overloads should work, but I'd like to hear some thoughts by actual users of the lib:
```ts
// 1. Old variant, should be deprecated or removed, depending on the bump that's going to be necessary for S2
// beginInclusion() will include with S0 or insecure
// beginInclusion(true) will include insecure
public async beginInclusion(includeNonSecure: boolean = false): Promise<boolean>

// 2. New variant, explicit insecure inclusion:
public async beginInclusion(security: Security.None): Promise<InclusionResult>

// 3. New variant, inclusion with S0 (or insecure, depending on the outcome):
public async beginInclusion(security: Security.S0): Promise<InclusionResult>

// 4. New variant, inclusion with S2. Node might be communicating with S2, S0 or insecure, depending on the user choice and outcome
public async beginInclusion(security: Security.S2, userCallbacks: /* t.b.d. */): Promise<InclusionResult>

// 5. LATER: inclusion with S2 SmartStart. Can include multiple nodes, requires the DSKs and security class choices to be done beforehand
// Nodes might be communicating with S2, S0 or insecure, depending on the user choice and outcome
public async beginInclusion(security: Security.S2_SmartStart, provisioningList: /* t.b.d. */): Promise<InclusionResult>
```

**Comment by marcus-j-davies** [comment]
Hey!

From the top of my head
If the process means the user must grant which Security Mode to accept, could we reduce this down to.

```
public async beginInclusion(security: Security.None):
public async beginInclusion(security: Security.SmartStart, provisioningList):
public async beginInclusion(security: Security.AsGranted, userCallback): /* <<------ decided as per user interaction */
```
essentially `AsGranted`
yields the security mode returned by the `userCallback`

`userCallback` will of course include what's supported by the device, and the security to be granted at that point?

>the user to choose which ones they want to grant. This includes S0, so a node included with S2 may communicate entirely with S0 afterwards.

Seems like as this includes S0 - is there a need to have an enum for both S2 an S0, when the prompt can include S0 as an option.

The `InclusionResult`, will include what was granted by the user, and what actually  took place (allowing the App to warn the user)

**Comment by marcus-j-davies** [comment]
To Extend on my Input:

if the device only supports S0, then the `userCallback` is not called (no other mode will work on the device) , but `InclusionResult` will naturally show the device has been included via S0

**Comment by AlCalzone** [comment]
To clarify: Inclusion with S0 works entirely different than inclusion with S2. Only S2 inclusion gives a device the chance to request security classes (including S0), but by that point the decision to use S0 or S2 for inclusion has already happened.

I guess a sensible modification/addition would be "as high as possible", which would use S2 if supported, falls back to S0 if that is supported and insecure otherwise. The `userCallback(s)` would have to be provided always but will only be invoked where applicable.

The `userCallback(s)` would be a set of functions the driver can call to tell the application to ask the user about security classes etc. and get the result, as well as notifying about the inclusion status/outcome.

**Comment by marcus-j-davies** [comment]
Ah ok, so maybe `None`, `SmartStart`, `AsHighAsPossible`.

And the callback, as you said, only taking place for S2.

I think that is where I was trying to go - not articulated well however 😬

If I have that correct! 🤷‍♂️

**Reply by AlCalzone** [reply]
We might need `S0` and `S2` as separate options for the cases where you absolutely don't want `S0` or unencrypted communication or where you absolutely want `S0` instead of `S2` - although I can't see why you'd want the latter.

**Reply by marcus-j-davies** [reply]
Maybe something like my previous.
```
public async beginInclusion(Security.None)
public async beginInclusion(Security.SmartStart, provisioningList)
public async beginInclusion([Security.S0, Security.S2], userCallback)
```

S2 will always be attempted over S0 in the first instance. This would allow the user to have strict control of not allowing a certain mode - by omitting it from the list, and `userCallback` only being consumed for S2.

This will also satisfy the `AsHighAsPossible` thing

Admittedly, this is just my 2c, Be interesting to read other opinions.

**Comment by blhoward2** [comment]
We generally preach for people to not include securely unless it's a cover. I know S2 isn't as bad for performance but I think I'd make the option not secure by default. 

Way back when on Hubitat I was presented with a list and had no idea if I should include securely or which to pick. That's really more on the UIs perhaps, but I think we need to explain the differences and have one labeled as the default.

**Reply by AlCalzone** [reply]
The Z-Wave Alliance is pushing for security hard, which is why all new devices MUST support S2. Of course experience will show how well it works, but there is almost no overhead besides some extra bytes when compared to no security.

I'd honestly advocate for S2 by default.

**Reply by blhoward2** [reply]
If you think it's not too big of a performance hit, then I defer to you honestly. I just think the actual choices displayed to the user must be understandable for lay people.

**Reply by marcus-j-davies** [reply]
The way I see this, is having 3 overloads for the developer.

No Security: `beginInclusion()` or `beginInclusion(Security.None)`
SmartStart:   `beginInclusion(Security.SmartStart, someList)`
Secure:         `beginInclusion([Security.S0, Security.S2], userCallback)`

The latter accepting a list of security Modes to be attempted (Highest security mode always take priority) The UI, can then show the first 2 modes as is, but for the 3rd - the UI can (optionally) ask the user, as to what security Modes they wish to attempt.

if S2 is used, a further dialog is displayed on receipt of the `userCallback`, during the 'handshake' process as to what S2 variant to grant -> Unauthorised, Authorised (DSK) etc etc.

the `InclusionResult` will always return as to what Mode was successful, to allow the rqeuirement to warn the user of a lower mode being used to what they wanted.

This is just the way I envision it, not sure if what I think (or how I imagine it working) is 'correct'. 🤷

**Reply by marcus-j-davies** [reply]
I suppose another question, is how to enforce the S2 default, when no callback has been provided.

Should it be trusted to the UI to enforce a default, or Driver side.
if Driver side, throw if no callback has been provided?

**Reply by robertsLando** [reply]
I agree with @marcus-j-davies but with a note

> No Security: beginInclusion(Security.None)
SmartStart: beginInclusion(Security.SmartStart, someList)
Secure: beginInclusion([Security.S0, Security.S2], userCallback)

`beginInclusion()` Should use S2 if available, no security if not

**Reply by marcus-j-davies** [reply]
I believe `beginInclusion()` will need to throw (for lack of callback), if we are having S2 seen as a default.

So the potential variants (I think) are

Invalid (Throw): `beginInclusion()`
No Security: `beginInclusion(Security.None)`
SmartStart: `beginInclusion(Security.SmartStart, someList)`
Highest Supported By Device: `beginInclusion([Security.None, Security.S0, Security.S2], userCallback)`
S2 or No Security : `beginInclusion(userCallback)` <--- Seen as the default.

Adding `Security.None` to the`Highest Supported By Device` allows devices to be included that don't offer any security, not adding `Security.None`, means, if the device does not support any security - then it is not included.


We better let @AlCalzone wade in here 😆

**Reply by AlCalzone** [reply]
> the actual choices displayed to the user must be understandable for lay people. 

Our documentation around this, including how the dialogs should look like and function, needs to be very specific and explain everything IMO. I think we should even go as far as giving recommendations for the UI and it's labels.

**Comment by AlCalzone** [comment]
Counter point to "as high as possible": you might end up with S0 if S2 isn't supported, which might not be ideal depending on the kind of device.

**Comment by AlCalzone** [comment, marked as answer]
In the first iteration I think I'm going with these inclusion strategies. Realistically, you could argue against all but Default and SmartStart, but I can already hear the call for more control.

**Default:**  
Prefer Security S2 if supported, use Security S0 if absolutely necessary, don't use encryption otherwise.
Issues a warning if Security S2/S0 is attempted to be used but secure bootstrapping fails.
👍🏻 **This is the recommended** strategy and should be used unless there is a good reason not to.

---

**SmartStart:**  
Include using SmartStart (requires Security S2).
Can't include a device that does not support Smart Start.
Issues a warning if secure bootstrapping fails.
👍🏻👍🏻 **Should be preferred** over **Default** if supported.

---

~~**HighestSupported:**  
Prefer the highest supported security mode that is supported by the device, otherwise include without encryption.
Issues a warning if Security S0 or S2 is supported, but the secure bootstrapping fails.
👎🏻 **Not recommended** because S0 should be used sparingly and you could end up with S0 on devices where it doesn't really make sense.~~

---

**Insecure:**  
Don't use encryption, even if supported.
👎🏻 **Not recommended**, because S2 should be used where possible.

---

**Security_S0:**  
Use Security S0, even if a higher security mode is supported.
Issues a warning if Security S0 is not supported or the secure bootstrapping fails.
👎🏻👎🏻 **Not recommended** because S0 should be used sparingly and S2 preferred whereever possible.

---

**Security_S2:**  
Use Security S2 and issue a warning if it is not supported or the secure bootstrapping fails.
🤷🏻‍♂️ Not really an advantage over **Default**. Less versatile, more complicated for the user.

---

Default, ~~HighestSupported~~ and Security_S2 will take user callbacks that allow interactively validating the DSK and choosing security classes.
SmartStart will take a provisioning list - but might end up needing no inclusion method at all (I need to read up more on this).
Insecure and Security_S0 won't need any arguments.

**Comment by marcus-j-davies** [comment]
That all seems intuitive!

I think **HighestSupported** will be a favourite by many users however, there are some devices that don't support S2, and I would imagine some users, still want that 'secure' feel.

The options do cater for all situations however! - nice work!

**Comment by kpine** [comment]
The **Default** strategy does not account for the numerous locks and garage door openers that only support S0. This means that applications would really end up having to recommend using the **HighestSupported** strategy for all locks and garage door openers, and **Default** for everything else, to avoid confusion regarding S2 vs S0. That's basically what is advised today, we tell users to enable security for locks and garage door openers, and the default (for both HA and zwavejs2mqtt) is non-secure inclusion.

Way back in 2018, Hubitat [announced](https://community.hubitat.com/t/hub-update-1-1-7/5215) that only locks and garage door openers would default to secure inclusion, and every other device would be included without security by default (not sure if that changed with S2 support). Is this something you could also do? This behavior sounds ideal to me, and is the same advice HA gives users today. A _minor_ tweak (I'm sure... 😄) to the **Default** strategy would cover the most common use cases:

1. If the device supports Security S2, include it with S2.
2. Else, if the device is a lock or garage door opener (are there other types?), include it with Security S0. These devices cannot properly function w/o security.
3. Else, if the user turned on a flag that forces Security S0, and S0 is supported, include it with S0.
4. Else, include it without security.

From the driver perspective, is this too complicated or not possible? It would certainly be a great usability improvement. Having to only click a single "Add" button for most cases is a better experience and easier to support.

**Reply by AlCalzone** [reply]
Good point. When we're adding a new node (not replacing a failed one), we do know the device class, so we can make that decision with very little effort. I think making the strategy you outlined the **Default** seems good.

**Reply by marcus-j-davies** [reply]
Hence why I think **HighestSupported**  will initially be a favorite include mode - for those not supporting S2 - but wont operate without Security.

> Else, if the device is a lock or garage door opener (are there other types?), include it with Security S0. These devices cannot properly function w/o security.

I think this might be risky/messy, is it of certainty that **ALL** these devices do support security, do specifications state they must - it may ? I don't actually know myself 🤷‍♂️.

**Reply by AlCalzone** [reply]
The specifications require encryption for certain device types, so this is easily configurable.

**Reply by marcus-j-davies** [reply]
Cool, I thought it was of an assumption as apposed to specification.

>  use Security S0 if absolutely necessary.

if the device type must use security as per spec?

**Reply by AlCalzone** [reply]
yep

**Comment by AlCalzone** [comment]
![grafik](https://user-images.githubusercontent.com/17641229/128861749-d52444f5-31c5-4168-a501-e20a9f8bf126.png)

🎉🎉🎉

**Reply by marcus-j-davies** [reply]
Smart Start! Smart Start! Smart Start! Smart Start!.

You sir are on Fire!!

**Comment by gdt** [comment]
I'm writing about a world with no security and S0, and not really trying to comment on the S2 issues.

I realize there is an efficiency argument to use no security rather than S0, but my strong preference is to have everything be encrypted, even if it's a lamp switch, a flood sensor, or a motion alarm, and even if this only addresses some concerns but not traffic analysis.  When there is an efficiency/security tradeoff it seems like a strange choice to choose efficiency by default.    So this feels like something that should perhaps have an explicit wizard, where people are asked "When adding devices that are not locks or doors, do you want "no security" (more efficient) or "S0 security" (more security, but less efficient" to set a toggle for that.

I'm still on the openzwave integration, about to migrate to zwavejs soon.  I have 13 devices, all included with S0, and not too much happens on my network, and it works fine, in terms of commands responding within a second and no battery life issues.   So I have zero motivation to move to cleartext to save bandwidth.

I also don't understand the thinking behind "S2 only".   It seems like it will be a long time before even 25% of zwavejs users will have only devices that support S2.

**Reply by marcus-j-davies** [reply]
> So this feels like something that should perhaps have an explicit wizard, where people are asked "When adding devices that are not locks or doors, do you want "no security" (more efficient) or "S0 security" (more security, but less efficient"

ZWave JS is the driver behind the  application/user interface that employs its magic, sure a wizard can be provided, but that is down to the user interface that employs the driver.

Many applications use ZWave JS.

I think there will be guidance, about what each mode provides, but  how that is displayed is ultimately down to the application, providing the information is consistent

**Comment by gdt** [comment]
A fair point about separation of UI, but there needs to be an option to prefer S0 to none so that can be selected.

I wonder if trying to have a single strategy token wraps up non-orthogonal decisions, and if it would be better to have several toggles for the policy choices.   I am guessing that "use S2 when available" is something almost everybody will want on, and then there's "if none and S0 are the feasible choices, choose [which]".   With that and requiring a "include this anyway even though it's a lock/cover and doesn't do S2" flag for including specific devices, that might be sufficient (or easily extendable to be sufficient).   I find a few "when faced with situation X do Y' questions easier to follow than N complete strategies (and my formal training is in CS).

**Comment by AlCalzone** [comment]
Implementation is up at https://github.com/zwave-js/node-zwave-js/pull/3170

@gdt What you describe is already incorporated in the default strategy. As it stands, that strategy...
* Uses S2 if possible
* If not, uses S0 if required by the device class OR opted in by the user
* uses no encryption otherwise

The individual **only S2** OR **only S0** OR **only Insecure** strategies are there for `replaceFailedNode` where we have no other choice and to avoid future feature requests.

**Reply by Ikcelaks** [reply]
The **only S0** and **only Insecure** strategies are also necessary for forcing identical security levels for two devices that support different levels for security, which I _believe_ is necessary for them to communicate directly through association groups.

Is that correct?

**Reply by AlCalzone** [reply]
If I understand correctly, this should work in one direction. A node can control other nodes with any of its granted security classes. However, nodes won't accept any commands on a lower security class than their highest one.

So let's say, Node A has AccessControl, Authenticated and S0, Node B has Authenticated and S0, Node C only S0.
Node A will be able to control Node B (using S2 Authenticated), Node B will be able to control Node C (using S0), but not vice versa.
